Readme.md
<p align="center">
  <img src="https://blauwfruit.nl/logos/logo.svg" width="300"/>
</p>


PrestaShop Module for AfterPay: Achteraf betalen
===============

## How to install the module?

* Download the latest version of the module (the '.zip' file) via the [Releases page](https://github.com/blauwfruit/afterpaynl/releases) which is compatible with PrestaShop 1.6 to 1.7;
* Go to the backoffice of your PrestaShop webshop;
* In your backoffice, go to 'Modules' and then 'Module Manager' and choose 'Upload a module';
* Click on 'select file' and upload the .zip file.

## Configuration

* Modules &rarr; Module Manager &rarr; Achteraf betalen met AfterPay
* After the module has been installed, click on 'Configure';
* Enter your 'Merchant ID', 'Portfolio ID' and 'Password' (provided by AfterPay). You should do this for every applicable account: B2C, B2B and B2C Belgium;
* If you have enabled the PrestaStore Multistore functionality, fill out these details per store;
* Click on 'Save'.

## Address format

To be absolutely sure that people won’t be rejected because of wrong address details, we strongly advise to change the address format:

* Localisation &rarr; Country &rarr; Netherlands; 
* Then, under 'Addres format', you can rearrange the order to: 
```
firstname lastname
company
vat_number
address1 address2
postcode city
Country:name
phone_mobile
phone
```
* Click on 'Save'.

`address2` is used as the house number for AfterPay. It is important to make it mandatory. Also, in PrestaShop 1.7, make sure you make the phone number obligatory. Otherwise the customer will be rejected: 

* Customers → Addresses → click on the button 'Set required fields for this section';
* Select `address2`; 
* Unselect `phone` and/or `phone_mobile`;
* Click on 'Save'.

CHANGELOG.txt
3.2.4

- style: Prepare module for publishing on GitHub

3.2.3

- fix: only load hookDisplayOrderConfirmation when payment is with (this) AfterPay module
- fix: use fetch function to display confirmation.tpl
- fix: use object $param['order'] for 1.7.x

3.2.2

- fix: add absolute path when fetching template inside a template (module:afterpaynl/views/templates/front/..) for compatability with 1.7.7.5

3.2.1

- Improved template fetching
- Improved logic to check if customer is eligible for AfterPay

3.2.0

- Merge 1.6 with 1.7
- Improved birtday validation
- Improved error messaging
- Displaying payment options always, warn client on AfterPay payment page
- Renames payment page to `pay`

3.1.1

- Improvement passing on phone number to the AfterPay API 

3.1.0

- Automatically making phone required as per the requirements of AfterPay
- Documentation extended


translations/nl.php
<?php

global $_MODULE;
$_MODULE = array();
$_MODULE['<{afterpaynl}prestashop>afterpaynl_4b1ba5acfa104a68f650d3890e4ab6af'] = 'Afterpay NL';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_94f3ca09f0de8fe41e912d6c9ffd3086'] = 'Reken direct af met Afterpay.';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_92d4773a9f32fc3333cb8238471cf763'] = 'Je moet de cURL module ingeschakeld hebben voor deze module';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_650be61892bf690026089544abbd9d26'] = 'Modus';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_79d7ceaf09e387e9b1966467b0b2ea56'] = 'Live?';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_73cbf76056a5f17df10c0f23c72ffb0a'] = 'Zet aan als je live wilt gaan';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_521c36a31c2762741cf0f8890cbe05e3'] = 'Aan';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_d15305d7a4e34e02489c74a5ef542f36'] = 'Uit';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_c9cc8cce247e49bae79f15173ce97354'] = 'Opslaan';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_787f3a718de9cf0adc436706e939f53f'] = 'Business to Consumer instellingen';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_229a7ec501323b94db7ff3157a7623c9'] = 'Merchant ID';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_e565311d40f8fad3b268b7749e2d0744'] = '2017000000';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_cfca908360088dac64f938efd6070b11'] = 'Portofolio ID';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_c4ea0aedee24fe7f7df564bef5599a62'] = 'Voorbeeld: 1';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_dc647eb65e6711e155375218212b3964'] = 'Wachtwoord';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_df28ce1d23cc41fca2f15b1a9793a13e'] = 'Voorbeeld: 8aka4a094b (laat leeg om uit te zetten)';
$_MODULE['<{afterpaynl}prestashop>afterpaynl_1466b3052d59b2f56a150a3100fad6f9'] = 'Business to Business instellingen';
$_MODULE['<{afterpaynl}prestashop>redirect_e5b1be1730651456abf9335164ef0c9e'] = 'Log eerst in.';
$_MODULE['<{afterpaynl}prestashop>redirect_da9544014256b8764c15fc4c8550ca9b'] = 'Het minimale order bedrag is %s';
$_MODULE['<{afterpaynl}prestashop>redirect_c780dfbd7ad85321362a5fac359f3606'] = 'Accepteer de algemene voorwaarden van AfterPay';
$_MODULE['<{afterpaynl}prestashop>redirect_03b76f719b04402404f722c2d96aef43'] = 'De data komt niet van een vertrouwde bron';
$_MODULE['<{afterpaynl}prestashop>redirect_480981c4a28c4f0f840f2afbc4d08393'] = 'Verjaardag niet opgegeven';
$_MODULE['<{afterpaynl}prestashop>redirect_b71ea5715f50beff8538c2cf978d4ab2'] = 'Geslacht niet opgegeven';
$_MODULE['<{afterpaynl}prestashop>redirect_d38fe54d243e3c96706917b9d946b6c0'] = 'Bedrijfsnaam is niet opgegeven';
$_MODULE['<{afterpaynl}prestashop>redirect_8f9922ac46212e2d3372d2928e2b1449'] = 'KVK-nummer niet opgegeven';
$_MODULE['<{afterpaynl}prestashop>redirect_31d6fa4c4fe84b9733a381428f4e5975'] = 'KVK-nummer niet correct';
$_MODULE['<{afterpaynl}prestashop>redirect_aac772216aecbeca0e86d06671fe985a'] = 'Geboortedag';
$_MODULE['<{afterpaynl}prestashop>redirect_019ec3132cdf8ee0f2e2a75cf5d3e459'] = 'Geslacht';
$_MODULE['<{afterpaynl}prestashop>redirect_127469a6b4253ebb77adccc0dd48461e'] = 'Man';
$_MODULE['<{afterpaynl}prestashop>redirect_773442a125403cbb194c24c5a8d59b88'] = 'Vrouw';
$_MODULE['<{afterpaynl}prestashop>redirect_c281f92b77ba329f692077d23636f5c9'] = 'Bedrijfsnaam';
$_MODULE['<{afterpaynl}prestashop>redirect_ea43b89e09914327b285455f603218e1'] = 'KVK-nummer';
$_MODULE['<{afterpaynl}prestashop>redirect_6e102731c5d2d9b7c10e31e3f4e02ed6'] = 'Ga akkoord met de algemene voorwaarden van AfterPay';
$_MODULE['<{afterpaynl}prestashop>redirect_4d15f1e78208656950fe55c5f2696939'] = 'CADEAUVERPAKKING';
$_MODULE['<{afterpaynl}prestashop>redirect_23277045d8271f587f10fd55c284324f'] = 'Cadeauverpakking';
$_MODULE['<{afterpaynl}prestashop>redirect_e92dc03a9834d3e1e0645e6bbc55ce18'] = 'VERZENDKOSTEN';
$_MODULE['<{afterpaynl}prestashop>redirect_a5e35abc0c9e2d2784d0ef619b36448b'] = 'Verzendkosten';
$_MODULE['<{afterpaynl}prestashop>redirect_d37fe1215db68c6abe92a81e11852711'] = 'AFRONDING';
$_MODULE['<{afterpaynl}prestashop>redirect_f73b877420a9122cfc16b01372371427'] = 'Afrondingsverschil';
$_MODULE['<{afterpaynl}prestashop>redirect_c453a4b8e8d98e82f35b67f433e3b4da'] = 'Betaling';
$_MODULE['<{afterpaynl}prestashop>redirect_902b0d55fddef6f8d651fe1035b7d4bd'] = 'Fout';
$_MODULE['<{afterpaynl}prestashop>redirect_3cd3e3ff14b421a10f1d0119d71e64d8'] = 'Aankoop afronden met AfterPay';
$_MODULE['<{afterpaynl}prestashop>configure_a82be0f551b8708bc08eb33cd9ded0cf'] = 'Informatie';
$_MODULE['<{afterpaynl}prestashop>afterpay_error_be7b7cb10b38caa3792dee263c2ae062'] = 'Oeps. Er is iets verkeerd gegaan.';
$_MODULE['<{afterpaynl}prestashop>afterpay_error_d22da7129b438d7e07b8f1816097d81a'] = 'Ga terug en corrigeer de nodige gegevens';
$_MODULE['<{afterpaynl}prestashop>error_a25c753ee3e4be15ec0daa5a40deb7b8'] = 'Er is een fout opgetreden';
$_MODULE['<{afterpaynl}prestashop>redirect_f2cf785dffc88e38702dc8a2232e5b10'] = 'Corrigeer het volgende';
$_MODULE['<{afterpaynl}prestashop>redirect_20484c83dcc00f01d509f8561a6243d1'] = 'Aankoop afronden';
$_MODULE['<{afterpaynl}prestashop>redirect_eb18dd00e86aff4b5b5d15e938382b49'] = 'Kies een andere betaalmethode';
$_MODULE['<{afterpaynl}prestashop>confirmation_2f47ce9aab1b2c9b6281b27d85061539'] = 'Bedankt voor  je order bij %s';
$_MODULE['<{afterpaynl}prestashop>confirmation_d76d6d0c72e116987c13bd29017778d7'] = 'Je order is succesvol geplaatst, voor vragen kan je refereren aan %s.';
$_MODULE['<{afterpaynl}prestashop>payment_467412470ffa7271d229ad4d3bcb5fd8'] = 'Het minimale bedrag benodigd voor AfterPay:';
$_MODULE['<{afterpaynl}prestashop>payment_82700323ccdccb68524185787af33206'] = 'Betaal met AfterPay';
$_MODULE['<{afterpaynl}prestashop>payment_1cfe4babda30cb80cefda2e43dba946f'] = 'Bestel nu, betaal later';
$_MODULE['<{afterpaynl}prestashop>payment_e7f6be0acb0828fe7f43d21a9f1d8266'] = 'Betaal AfterPay';
$_MODULE['<{afterpaynl}prestashop>payment_96efc9700576d4e18a6ca2f7ca1682da'] = '(de bestelling wordt direct verwerkt, de betaling kunt u later voldoen)';


afterpaynl.php
<?php
/**
*   AfterPay Netherlands
*
*   Do not copy, modify or distribute this document in any form.
*
*   @author     Matthijs <matthijs@blauwfruit.nl>
*   @copyright  Copyright (c) 2013-2021 blauwfruit (https://blauwfruit.nl)
*   @license    Proprietary Software
*
*/

if (!defined('_PS_VERSION_')) {
    exit;
}

class Afterpaynl extends PaymentModule
{
    public $notEligibleReason = array();

    public function __construct()
    {
        $this->name = 'afterpaynl';
        $this->tab = 'payments_gateways';
        $this->version = '3.2.4';
        $this->author = 'blauwfruit';
        $this->need_instance = 0;
        $this->bootstrap = true;
        $this->module_key = '409341bee0baf049a52faff643b4f905';
        $this->controllers = array('pay');
        parent::__construct();
        $this->displayName = $this->l('Afterpay NL');
        $this->description = $this->l('Pay directly with Afterpay.');
        $this->limited_countries = array('NL', 'BE');
        $this->limited_currencies = array('EUR', 'EU', 'EURO');
        $this->ps_versions_compliancy = array('min' => '1.6', 'max' => _PS_VERSION_);
    }

    /**
     */
    public function install()
    {
        if (extension_loaded('curl') == false) {
            $this->_errors[] = $this->l('You have to enable the cURL extension on your server to install this module');
            return false;
        }
        
        $sql = "SELECT * FROM "._DB_PREFIX_."required_field WHERE object_name = 'Address' AND field_name = 'phone'";
        if (!Db::getInstance()->getRow($sql)) {
            Db::getInstance()->insert('required_field', array(
                'object_name' => 'Address',
                'field_name' => 'phone',
            ));
        }

        $sql = "SELECT * FROM "._DB_PREFIX_."required_field WHERE object_name = 'CustomerAddress' AND field_name = 'phone'";
        if (!Db::getInstance()->getRow($sql)) {
            Db::getInstance()->insert('required_field', array(
                'object_name' => 'CustomerAddress',
                'field_name' => 'phone',
            ));
        }

        return parent::install() &&
            $this->registerHook('header') &&
            $this->registerHook('paymentOptions') &&
            $this->registerHook('paymentReturn') &&
            $this->registerHook('displayOrderConfirmation') &&
            $this->registerHook('displayPayment');
    }

    public function uninstall()
    {
        return parent::uninstall();
    }

    /**
     * Load the configuration form
     */
    public function getContent()
    {
        $html = '';
        if (Tools::isSubmit('submit'.$this->name)) {
            foreach (array('AFTERPAYNL_MODE', 'AFTERPAYNL_MERCHANT_ID_B2C', 'AFTERPAYNL_PORTOFOLIO_ID_B2C',
                'AFTERPAYNL_PASSWORD_B2C', 'AFTERPAYNL_B2C_MAXIMUM_FIRST_AMOUNT',
                'AFTERPAYNL_B2C_MAXIMUM_SECOND_AMOUNT',
                'AFTERPAYNL_MERCHANT_ID_B2B', 'AFTERPAYNL_PORTOFOLIO_ID_B2B',
                'AFTERPAYNL_PASSWORD_B2B', 'AFTERPAYNL_B2B_MAXIMUM_FIRST_AMOUNT',
                'AFTERPAYNL_B2B_MAXIMUM_SECOND_AMOUNT',

                'AFTERPAYNL_MERCHANT_ID_B2C_BE', 'AFTERPAYNL_PORTOFOLIO_ID_B2C_BE',
                'AFTERPAYNL_PASSWORD_B2C_BE', 'AFTERPAYNL_B2C_BE_MAXIMUM_FIRST_AMOUNT',
                'AFTERPAYNL_B2C_BE_MAXIMUM_SECOND_AMOUNT'



                ) as $ConfigName) {
                    Configuration::updateValue($ConfigName, Tools::getValue($ConfigName));
            }
        }
        $html .= $this->renderForm();
        $this->context->smarty->assign(array(
            'token' => sha1(_PS_BASE_URL_),
            'shop_url' => _PS_BASE_URL_.__PS_BASE_URI__,
            'context' => $this->context
        ));
        $output = $this->context->smarty->fetch($this->local_path.'views/templates/admin/configure.tpl');
        return $html.$output;
    }


    /**
     * Create the structure of your form
     */
    public function renderForm()
    {
        $mode = array(
            'form' => array(
                'legend' => array(
                    'title' => $this->l('Mode'),
                    'icon' => 'icon-cogs'
                ),
                'input' => array(
                    array(
                        'type' => 'switch',
                        'name' => 'AFTERPAYNL_MODE',
                        'label' => $this->l('Live?'),
                        'desc' => $this->l('Turn on if you want to push it live'),
                        'values' => array(
                            array(
                                'id' => 'on',
                                'value' => 1,
                                'name' => $this->l('On'),
                            ),
                            array(
                                'id' => 'off',
                                'value' => 0,
                                'name' => $this->l('Off'),
                            ),
                        ),
                    ),
                ),
                'submit' => array(
                    'title' => $this->l('Save'),
                    'class' => 'btn btn-default pull-right'),
            )
        );
        $B2C = array(
            'form' => array(
                'legend' => array(
                    'title' => $this->l('Business to Consumer settings'),
                    'icon' => 'icon-cogs'
               ),
                'input' => array(
                    array(
                      'type'     => 'text',
                      'label'    => $this->l('Merchant ID'),
                      'name'     => 'AFTERPAYNL_MERCHANT_ID_B2C',
                      'size'     => 50,
                      'required' => true,
                      'desc'     => $this->l('2017000000')
                   ),
                    array(
                      'type'     => 'text',
                      'label'    => $this->l('Portofolio ID'),
                      'name'     => 'AFTERPAYNL_PORTOFOLIO_ID_B2C',
                      'size'     => 50,
                      'required' => true,
                      'desc'     => $this->l('Example: 1')
                   ),
                    array(
                      'type'     => 'text',
                      'label'    => $this->l('Password'),
                      'name'     => 'AFTERPAYNL_PASSWORD_B2C',
                      'size'     => 50,
                      'required' => true,
                      'desc'     => $this->l('Example: 8aka4a094b (leave empty to disable)')
                   ),
               ),
            'submit' => array(
                'title' => $this->l('Save'),
                'class' => 'btn btn-default pull-right'),
           ),
        );
        $B2B = array(
            'form' => array(
                'legend' => array(
                    'title' => $this->l('Business to Business settings'),
                    'icon' => 'icon-cogs',
                    'descr' => 'Hello',
               ),
                'input' => array(
                    array(
                      'type'     => 'text',
                      'placeholder' => 'Merchant ID',
                      'label'    => $this->l('Merchant ID'),
                      'name'     => 'AFTERPAYNL_MERCHANT_ID_B2B',
                      'size'     => 50,
                      'required' => true,
                      'desc'     => $this->l('2017000000')
                   ),
                    array(
                      'type'     => 'text',
                      'placeholder' => 'Portofolio ID',
                      'label'    => $this->l('Portofolio ID'),
                      'name'     => 'AFTERPAYNL_PORTOFOLIO_ID_B2B',
                      'size'     => 50,
                      'required' => true,
                      'desc'     => $this->l('Example: 1')
                   ),
                    array(
                      'type'     => 'text',
                      'placeholder' => 'Password',
                      'label'    => $this->l('Password'),
                      'name'     => 'AFTERPAYNL_PASSWORD_B2B',
                      'size'     => 50,
                      'required' => true,
                      'desc'     => $this->l('Example: 8aka4a094b (leave empty to disable)')
                   ),
               ),
            'submit' => array(
                'title' => $this->l('Save'),
                'class' => 'btn btn-default pull-right')
           ),
        );
        $B2C_BE = array(
            'form' => array(
                'legend' => array(
                    'title' => $this->l('Business to Consumer for Belgium settings'),
                    'icon' => 'icon-cogs',
                    'descr' => 'Hello',
               ),
                'input' => array(
                    array(
                      'type'     => 'text',
                      'placeholder' => 'Merchant ID',
                      'label'    => $this->l('Merchant ID'),
                      'name'     => 'AFTERPAYNL_MERCHANT_ID_B2C_BE',
                      'size'     => 50,
                      'required' => true,
                      'desc'     => $this->l('2017000000')
                   ),
                    array(
                      'type'     => 'text',
                      'placeholder' => 'Portofolio ID',
                      'label'    => $this->l('Portofolio ID'),
                      'name'     => 'AFTERPAYNL_PORTOFOLIO_ID_B2C_BE',
                      'size'     => 50,
                      'required' => true,
                      'desc'     => $this->l('Example: 1')
                   ),
                    array(
                      'type'     => 'text',
                      'placeholder' => 'Password',
                      'label'    => $this->l('Password'),
                      'name'     => 'AFTERPAYNL_PASSWORD_B2C_BE',
                      'size'     => 50,
                      'required' => true,
                      'desc'     => $this->l('Example: 8aka4a094b (leave empty to disable)')
                   ),
               ),
            'submit' => array(
                'title' => $this->l('Save'),
                'class' => 'btn btn-default pull-right')
           ),
        );


        $helper = new HelperForm();
        $helper->submit_action = 'submit'.$this->name;
        $helper->module = $this;
        $helper->name_controller = $this->name;
        $helper->token = Tools::getAdminTokenLite('AdminModules');
        $helper->currentIndex = AdminController::$currentIndex.'&configure='.$this->name;
        $helper->title = $this->displayName;
        # B2C
        $helper->fields_value['AFTERPAYNL_MODE'] = Configuration::get('AFTERPAYNL_MODE');
        $helper->fields_value['AFTERPAYNL_MERCHANT_ID_B2C'] = Configuration::get('AFTERPAYNL_MERCHANT_ID_B2C');
        $helper->fields_value['AFTERPAYNL_PORTOFOLIO_ID_B2C'] = Configuration::get('AFTERPAYNL_PORTOFOLIO_ID_B2C');
        $helper->fields_value['AFTERPAYNL_PASSWORD_B2C'] = Configuration::get('AFTERPAYNL_PASSWORD_B2C');
        $AFTERPAYNL_B2C_MAXIMUM_FIRST_AMOUNT = Configuration::get('AFTERPAYNL_B2C_MAXIMUM_FIRST_AMOUNT');
        $helper->fields_value['AFTERPAYNL_B2C_MAXIMUM_FIRST_AMOUNT'] = $AFTERPAYNL_B2C_MAXIMUM_FIRST_AMOUNT;
        $AFTERPAYNL_B2C_MAXIMUM_SECOND_AMOUNT = Configuration::get('AFTERPAYNL_B2C_MAXIMUM_SECOND_AMOUNT');
        $helper->fields_value['AFTERPAYNL_B2C_MAXIMUM_SECOND_AMOUNT'] = $AFTERPAYNL_B2C_MAXIMUM_SECOND_AMOUNT;
        # B2B
        $helper->fields_value['AFTERPAYNL_MERCHANT_ID_B2B'] = Configuration::get('AFTERPAYNL_MERCHANT_ID_B2B');
        $helper->fields_value['AFTERPAYNL_PORTOFOLIO_ID_B2B'] = Configuration::get('AFTERPAYNL_PORTOFOLIO_ID_B2B');
        $helper->fields_value['AFTERPAYNL_PASSWORD_B2B'] = Configuration::get('AFTERPAYNL_PASSWORD_B2B');
        $AFTERPAYNL_B2B_MAXIMUM_FIRST_AMOUNT = Configuration::get('AFTERPAYNL_B2B_MAXIMUM_FIRST_AMOUNT');
        $helper->fields_value['AFTERPAYNL_B2B_MAXIMUM_FIRST_AMOUNT'] = $AFTERPAYNL_B2B_MAXIMUM_FIRST_AMOUNT;
        $AFTERPAYNL_B2B_MAXIMUM_SECOND_AMOUNT = Configuration::get('AFTERPAYNL_B2B_MAXIMUM_SECOND_AMOUNT');
        $helper->fields_value['AFTERPAYNL_B2B_MAXIMUM_SECOND_AMOUNT'] = $AFTERPAYNL_B2B_MAXIMUM_SECOND_AMOUNT;
        # B2C BE
        $helper->fields_value['AFTERPAYNL_MERCHANT_ID_B2C_BE'] = Configuration::get('AFTERPAYNL_MERCHANT_ID_B2C_BE');
        $AFTERPAYNL_PORTOFOLIO_ID_B2C_BE = Configuration::get('AFTERPAYNL_PORTOFOLIO_ID_B2C_BE');
        $helper->fields_value['AFTERPAYNL_PORTOFOLIO_ID_B2C_BE'] = $AFTERPAYNL_PORTOFOLIO_ID_B2C_BE;
        $helper->fields_value['AFTERPAYNL_PASSWORD_B2C_BE'] = Configuration::get('AFTERPAYNL_PASSWORD_B2C_BE');
        $AFTERPAYNL_B2C_BE_MAXIMUM_FIRST_AMOUNT = Configuration::get('AFTERPAYNL_B2C_BE_MAXIMUM_FIRST_AMOUNT');
        $helper->fields_value['AFTERPAYNL_B2C_BE_MAXIMUM_FIRST_AMOUNT'] = $AFTERPAYNL_B2C_BE_MAXIMUM_FIRST_AMOUNT;
        $AFTERPAYNL_B2C_BE_MAXIMUM_SECOND_AMOUNT = Configuration::get('AFTERPAYNL_B2C_BE_MAXIMUM_SECOND_AMOUNT');
        $helper->fields_value['AFTERPAYNL_B2C_BE_MAXIMUM_SECOND_AMOUNT'] = $AFTERPAYNL_B2C_BE_MAXIMUM_SECOND_AMOUNT;

        return $helper->generateForm(array($mode,$B2C,$B2B,$B2C_BE));
    }
    
    /**
     * Some styles to be loaded here
     */
    public function hookHeader()
    {
    }

    public function isEligible($companyName, $isoCode)
    {
        if ($companyName == '' && $isoCode == 'NL' && Configuration::get('AFTERPAYNL_PASSWORD_B2C') == '') {
            $this->notEligibleReason[] = $this->l('AfterPay is not available in your situation because you are a consumer residing in the Netherlands. Change the invoice and/or delivery address');
            return false;
        } elseif ($companyName !== '' && $isoCode == 'NL' && Configuration::get('AFTERPAYNL_PASSWORD_B2B') == '') {
            $this->notEligibleReason[] = $this->l('AfterPay is not available in your situation because you are a business residing in the Netherlands. Change the invoice and/or delivery address');
            return false;
        } elseif ($companyName == '' && $isoCode == 'BE' && Configuration::get('AFTERPAYNL_PASSWORD_B2C_BE') == '') {
            $this->notEligibleReason[] = $this->l('AfterPay is not available in your situation because you are a consumer residing in Belgium. Change the invoice and/or delivery address');
            return false;
        } elseif ($companyName !== '' && $isoCode == 'BE') {
            $this->notEligibleReason[] = $this->l('AfterPay is not available in your situation because you are a business residing in Belgium. Change the invoice and/or delivery address');
            return false;
        }

        return true;
    }

    public function getNotEligibleReason()
    {
        return $this->notEligibleReason;
    }

    /**
     * Return AfterPay in the order confirmation
     * @param (object) $params (order, etc.)
     * @return Display confirmation.tpl
     */
    public function hookDisplayOrderConfirmation($params)
    {
        if ((int) Tools::getValue('id_module') !== (int) $this->id) {
            return;
        }

        if ($this->isPS17()) {
            $this->smarty->assign('reference', $params['order']->reference);
            return $this->context->smarty->fetch('module:afterpaynl/views/templates/hook/confirmation.tpl');
        } else {
            $this->smarty->assign('reference', $params['objOrder']->reference);
            return $this->display($this->_path, 'views/templates/hook/confirmation.tpl');
        }
    }

    /**
     * Payment options for 1.7
     * @return (array) payment options
     */
    public function hookPaymentOptions($params)
    {
        $newOption = new PrestaShop\PrestaShop\Core\Payment\PaymentOption();
        $newOption->setModuleName($this->name)
                ->setCallToActionText($this->trans('Pay with AfterPay', array(), 'Modules.AfterPayNL.Front'))
                ->setAction($this->context->link->getModuleLink($this->name, 'pay', array()))
                ->setAdditionalInformation($this->fetch('module:afterpaynl/views/templates/front/payment_infos.tpl'))
                ->setLogo(Media::getMediaPath(_PS_MODULE_DIR_.$this->name.'/views/img/Logo-afterpay-checkout_L.png'));

        return array($newOption);
    }

    /**
     * Return AfterPay in the payment options for 1.6
     * @param (object) $params (cart, etc.)
     * @return payment.tpl
     */
    public function hookDisplayPayment($params)
    {
        $address_delivery = new Address((int) $params['cart']->id_customer);
        $country_delivery = new Country((int) $address_delivery->id_country);

        if (!$this->isEligible($address_delivery->company, $country_delivery->iso_code)) {
            return;
        }

        $this->smarty->assign(array(
            'module_dir' => $this->_path,
            'payment_url' => $this->context->link->getModuleLink('afterpaynl', 'pay', array(), true)
        ));
        return $this->display(__FILE__, 'views/templates/hook/payment.tpl');
    }    

    /**
     *  Payment return
     */
    public function hookPaymentReturn($params)
    {
        if (!$this->active) {
            return;
        }

        $state = $params['objOrder']->getCurrentState();
        if (in_array($state, array(
            Configuration::get('PS_OS_CHEQUE'),
            Configuration::get('PS_OS_OUTOFSTOCK'),
            Configuration::get('PS_OS_OUTOFSTOCK_UNPAID')))) {
            $this->smarty->assign(array(
                'total_to_pay' => Tools::displayPrice(
                    $params['order']->getOrdersTotalPaid(),
                    new Currency($params['order']->id_currency),
                    false
                ),
                'shop_name' => $this->context->shop->name,
                'checkName' => $this->checkName,
                'checkAddress' => Tools::nl2br($this->address),
                'status' => 'ok',
                'id_order' => $params['order']->id
            ));
            if (isset($params['order']->reference) && !empty($params['order']->reference)) {
                $this->smarty->assign('reference', $params['order']->reference);
            }
        } else {
            $this->smarty->assign('status', 'failed');
        }
        if ($this->isPS17()) {
            $this->display('module:afterpaynl/views/templates/hook/payment_return.tpl');
        } else {
            return $this->display($this->_path, 'views/templates/hook/payment_return.tpl');
        }
    }

    /**
     *  Is it PrestaShop 1.7?
     *  @return  bool 
     */
    public function isPS17()
    {
        return version_compare(_PS_VERSION_, '1.7', '>=');
    }
}
